File: apps/web/src/lib/portfolio.ts
import type { AutoRun } from "@/lib/auto-trader";

/**
 * Combine several single-asset bot runs into an equal-weight portfolio,
 * rebalanced per bar (mirror of apps/api backtest/portfolio.py). Diversifying
 * across assets lowers volatility and drawdown vs. any single asset.
 */
export interface PortfolioPoint {
  index: number;
  equity: number;
}

export interface PortfolioRun {
  curve: PortfolioPoint[];
  totalReturnPct: number;
  maxDrawdownPct: number;
  perAsset: { symbol: string; returnPct: number }[];
}

function returns(curve: { equity: number }[]): number[] {
  return curve.slice(1).map((p, i) => p.equity / curve[i]!.equity - 1);
}

export function combinePortfolio(runs: { symbol: string; run: AutoRun }[], startCash: number): PortfolioRun {
  if (runs.length === 0) {
    return { curve: [{ index: 0, equity: startCash }], totalReturnPct: 0, maxDrawdownPct: 0, perAsset: [] };
  }
  const perReturns = runs.map((r) => returns(r.run.steps.map((s) => ({ equity: s.equity }))));
  const horizon = Math.min(...perReturns.map((r) => r.length));
  const w = 1 / runs.length;

  let equity = startCash;
  let peak = startCash;
  let maxDd = 0;
  const curve: PortfolioPoint[] = [{ index: 0, equity }];
  for (let t = 0; t < horizon; t++) {
    const bar = perReturns.reduce((a, r) => a + w * r[t]!, 0);
    equity *= 1 + bar;
    peak = Math.max(peak, equity);
    maxDd = Math.max(maxDd, (peak - equity) / peak);
    curve.push({ index: t + 1, equity });
  }
  return {
    curve,
    totalReturnPct: (equity / startCash - 1) * 100,
    maxDrawdownPct: maxDd * 100,
    perAsset: runs.map((r) => ({ symbol: r.symbol, returnPct: r.run.summary.totalReturnPct })),
  };
}

File: apps/api/src/backtest/portfolio.py
"""Multi-asset portfolio backtest.

Running the strategy across several assets and combining the per-bar returns
diversifies idiosyncratic risk: the portfolio's volatility (and drawdown) is
typically lower than the average single asset, lifting risk-adjusted return.
Each asset is backtested independently with the same walk-forward + cost model;
returns are combined at equal (or supplied) weights, rebalanced each bar.
"""

from __future__ import annotations

import statistics
from dataclasses import dataclass
from typing import Mapping, Sequence

from src.backtest.engine import BacktestResult


@dataclass(frozen=True)
class PortfolioResult:
    total_return: float
    sharpe: float
    max_drawdown: float
    n_bars: int
    equity_curve: list[float]
    per_asset_return: dict[str, float]


def _returns(curve: Sequence[float]) -> list[float]:
    return [curve[i] / curve[i - 1] - 1.0 for i in range(1, len(curve))]


def combine_portfolio(
    results: Mapping[str, BacktestResult], weights: Mapping[str, float] | None = None
) -> PortfolioResult:
    """Combine per-asset backtests into an equal- (or weighted-) weight portfolio."""
    if not results:
        raise ValueError("no asset results to combine")
    names = list(results)
    w = {n: (weights[n] if weights else 1.0 / len(names)) for n in names}
    total_w = sum(w.values()) or 1.0

    per_asset_returns = {n: _returns(results[n].equity_curve) for n in names}
    horizon = min(len(r) for r in per_asset_returns.values())
    if horizon == 0:
        raise ValueError("assets have no overlapping return history")

    equity = 1.0
    curve = [1.0]
    port_returns: list[float] = []
    for t in range(horizon):
        bar = sum(w[n] / total_w * per_asset_returns[n][t] for n in names)
        equity *= 1.0 + bar
        curve.append(equity)
        port_returns.append(bar)

    peak = 1.0
    max_dd = 0.0
    for v in curve:
        peak = max(peak, v)
        max_dd = max(max_dd, (peak - v) / peak)
    sigma = statistics.stdev(port_returns) if len(port_returns) > 1 else 0.0
    sharpe = (statistics.fmean(port_returns) / sigma) if sigma > 0 else 0.0

    return PortfolioResult(
        total_return=equity - 1.0,
        sharpe=sharpe,
        max_drawdown=max_dd,
        n_bars=horizon,
        equity_curve=curve,
        per_asset_return={n: results[n].total_return for n in names},
    )
